### DISCUSSION TITLE
How can I use `AXUIElementCopyAttributValues()` with Avalonia?

### DISCUSSION THREAD

**Discussion by Irasil** [AvaloniaUI/Avalonia#13885] [Q&A]
I know this is not explicitly related to Avalonia, but I've had great experiences with the community's help, and I'm really desperate.

The app I'm developing needs to read window titles from other processes and react to them. I initially achieved this using `CGWindowListCopyWindowInfo` from `CoreGraphics.dylib`. However, to read the window title (`kCGWindowName`), it requires Screen Recording Permission. Since this permission allows screen recording and microphone access, I can't use it as customers will never accept it.

My current attempt involves using the Accessibility API of MacOS. The methods I need are `AXUIElementCreateApplication` and `AXUIElementCopyAttributValues()`. I obtain them through P/Invoke, but the program crashes immediately upon calling `AXUIElementCopyAttributValues()` with the error:

```txt
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000746954584180
Exception Codes:       0x0000000000000001, 0x0000746954584180
```

Does anyone know a way to test if the element I get from `AXUIElementCreateApplication` and the one the pointer points to is valid?

With Swift, it was no problem at all, and when I tried it with Objective-C (for a custom .dylib), I had to disable the sandbox. If I understand correctly, with an Avalonia app, I can only add entitlements after creating a .app bundle, is that correct? Is there a way to disable the sandbox during development, or is it already disabled by default? Since I was unsure, I tested it with a simple .Net app, but without success.

Or is there a completely different approach to read window titles from other processes?

[Here's](https://stackoverflow.com/questions/77621079/exc-bad-access-sigsegv-error-when-accessing-axuielementcopyattributvalues-in) the link to my Stack Overflow question, which includes more details about what I've tried.

I am very grateful for any help!

**Comment by thevortexcloud** [comment]
>  and the one the pointer points to is valid?

Well I assume a pointer value of 0 is invalid. Also because a lot of native programming languages don't have things like exceptions. you typically need to either check the return code of a method, or call some method to check for an error.

That said if I were to guess what the problem is: it's possibly related to how you compiled your native library. Any native library you compile needs to be usable from regular C/C++ to be usable from .NET. Without knowing much about Objective C, it _sounds_ like it's not directly compatible with either C/C++ out of the box. 
>Or is there a completely different approach to read window titles from other processes?

Can't you just use this?

https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.mainwindowtitle?view=net-8.0

You can find a list of processes using this:
https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.getprocesses?view=net-8.0

**Reply by Irasil** [reply]
Thanks for your response. I will take a closer look at compiling native libraries.

Using the `AXUIElement.h` API collection would be the primary goal, as it provides the possibility of using observers.

I initially tried it with Process.MainWindowTitle, but I always got empty strings. If I understood [this](https://github.com/dotnet/runtime/issues/70014#issuecomment-1164321879) comment correctly, this property is only implemented in Windows and therefore not an option.

I found [this](https://github.com/xamarin/xamarin-macios/issues/7536#issuecomment-1450384543) comment and will find out more about bindings and, if possible, create a binding myself. If I understand this correctly, that's exactly what I need.

**Reply by thevortexcloud** [reply]
> If I understood [this](https://github.com/dotnet/runtime/issues/70014#issuecomment-1164321879) comment correctly, this property is only implemented in Windows and therefore not an option.

Typical.

I think you can sort of do this using the Mac TFM? It looks like you can get a list of all running apps and their localised names.

https://learn.microsoft.com/en-us/dotnet/api/appkit.nsworkspace.runningapplications?view=xamarin-mac-sdk-14#appkit-nsworkspace-runningapplications

Which I assume is probably just the app itself rather than the window name. But it might return the active window name.

>I found https://github.com/xamarin/xamarin-macios/issues/7536#issuecomment-1450384543 comment and will find out more about bindings and, if possible, create a binding myself

Most of that comment just seems to be instructions on how to add a binding to that library. If it really is a proper C API all you need to do is import the methods from the correct library and then dispose of everything when you are done if it needs to be disposed of. In most cases you can just pass `IntPtr` around in your C# code if you don't need all the structs and what not. Some native APIs also require you to set them up in a specific way (EG call some init function) before you do anything else.

Also while this is not directly helpful for you, it looks like there are Rust bindings for that API. Which means you could also write your own wrapper for it in Rust and then compile that in a way that is compatible with C# P/Invoke. Which may be a lot simpler depending on what you need it to do. You could also probably do this with some other system language that is compatible with C. But this is probably be the last thing you should try.

**Reply by Irasil** [reply]
With `NSRunningApplication`, I have also tried it, and as you have already noticed, it only gives me the localizedName, which is the name of the app.

Could it be that I need to bind the Native Library to .Net (Xamarin.Mac) as shown in [this](https://www.youtube.com/watch?v=jd3RKcorq3o&t) or [this](https://www.youtube.com/watch?v=Ir5P2hdvWhg) video, or is that a completely wrong approach?

It's really silly, but I don't even know how to determine in which language the ApplicationService/HiService/AXUIElement.h framework is written.

> If it really is a proper C API all you need to do is import the methods from the correct library and then dispose of everything when you are done if it needs to be disposed of. In most cases you can just pass IntPtr around in your C# code if you don't need all the structs and what not.

Isn't that the approach I tried with P/Invoke? The problem is that I can't find any library on my Mac that I could include. From what I've read, `AXUIElementCreateApplication()` is part of `HiServices`, and this is in turn part of the `ApplicationServices` framework. If I look under `/System/Library/Frameworks/ApplicationServices.framework/..`, I can't find any .dylib file that I could include. That's why I created my own .dylib with the methods I needed, but even then, there was the same memory error. Since there was no Entry.Point Fail at `/System/Library/Frameworks/ApplicationServices.framework/Versions/Current/ApplicationServices`, I assumed that this works even though the ApplicationServices folder does not exist.

> Also while this is not directly helpful for you, it looks like there are Rust bindings for that API. Which means you could also write your own wrapper for it in Rust and then compile that in a way that is compatible with C# P/Invoke. Which may be a lot simpler depending on what you need it to do

I thank you for your time and this suggestion; I will look into it because I don't have any ideas left.

**Reply by thevortexcloud** [reply]
> Could it be that I need to bind the Native Library to .Net (Xamarin.Mac) as shown in [this](https://www.youtube.com/watch?v=jd3RKcorq3o&t) or [this](https://www.youtube.com/watch?v=Ir5P2hdvWhg) video, or is that a completely wrong approach?

I am not sure. I don't have any experience with programming for Macs sadly.

>It's really silly, but I don't even know how to determine in which language the ApplicationService/HiService/AXUIElement.h framework is written.

`.h` files are typically C/C++ header files. They are also one of my least favourite parts of C and C++ since it means the code  is spread across multiple files and can be painful to follow if you don't know C/C++ very well.

>  The problem is that I can't find any library on my Mac that I could include.

This is also a fairly common problem with dealing with using native libraries outside their intended environment. On Linux you can typically work it out from the man page. EG this is part of the man page for `fsync`

```
LIBRARY
       Standard C library (libc, -lc)

SYNOPSIS
       #include <unistd.h>

       int fsync(int fd);

       int fdatasync(int fd);

   Feature Test Macro Requirements for glibc (see feature_test_macros(7)):

       fsync():
           glibc 2.16 and later:
               No feature test macros need be defined
           glibc up to and including 2.15:
               _BSD_SOURCE || _XOPEN_SOURCE
                   || /* Since glibc 2.8: */ _POSIX_C_SOURCE >= 200112L

```

Which clearly tells you that you need to include `unistd.h` in your C code and it's part of `libc` and what versions the API is available in. But not all API manpages contain such information. The good ones do.

> Isn't that the approach I tried with P/Invoke? 

Pretty much.

>If I look under /System/Library/Frameworks/ApplicationServices.framework/.., I can't find any .dylib file that I could include

Strange. Apple's own documentation says _something_ should be there.

https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/OSX_Technology_Overview/SystemFrameworks/SystemFrameworks.html

`HIServices.framework` is also listed as a separate framework which means it's probably in a different library. Which also clearly says it's a C library: "These frameworks provide C-based interfaces ".

If I were to guess: You are actually missing the development headers. These are generally not included on a lot of systems (don't know about Mac specifically) since most people don't need them unless they are doing native programming (you technically don't need them in C# if you have the actual compiled binary already). I would guess installing XCode would probably install them for you. But I am not sure. On Linux you typically just have to install the `-dev` package for what ever it is you are trying to do. But I am not seeing much information about how to get them on a Mac.

So I guess try installing XCode/XCode CLI and see what it puts into the SDK folder?

**Reply by Irasil** [reply]
Thank you for all this informations.

XCode and Command Line Tools were already installed, but I didn't think to check the SDKs for the frameworks, as I've always used `/System/Library/Frameworks` for the other P/Invoke tasks I've done.

I found out that the contents of the folders are not identical.

`/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/HIServices.framework`

![Normal](https://github.com/AvaloniaUI/Avalonia/assets/91846030/4b5647f0-1d7f-4ff0-9413-0db1d9189403)

Here, I have no headers but instead have resources, which include the BridgeSupport folder. Inside, there are only 2 `.bridgesupport` files. I was actually expecting the `.dylib` here.

`/Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/HIServices.framework`

![sdk](https://github.com/AvaloniaUI/Avalonia/assets/91846030/15f5bcba-7949-4141-8994-38bca2520e11)


Here, I have no resources but the Header folder, which is full of `.h` and `.r` files.

![header](https://github.com/AvaloniaUI/Avalonia/assets/91846030/85fe7935-2cae-4bbc-b1d4-6ea89c143632)

From what I understand, the `.h` files don't really help me. What I need is a `.dylib`...
Because no matter which path I specify, there is either the error that no .dylib can be found or the memory error.

> Strange. Apple's own documentation says something should be there.

What I meant is that there are no `.dylib` files here. Other files are present.

What I don't understand is that when I use `DllImport /System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/HIServices.framework/Version/A/HIServices`, I get an error that HIServices.dylib doesn't exist. But if I use `/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices`, I get a memory error. However, there is neither `ApplicationServices.dylib` nor `HIService.dylib` in those folders.

I will look into the Rust binding (did you mean [eiz/accessibility](https://github.com/eiz/accessibility))?, create my own binding, and then try to create my own C++ .dylib containing the two methods I need.

As soon as I have new information I will report back
